File: HW16/HW16.cpp
#include <stdio.h>
#include <cmath>
#include <utility>
using namespace std;

pair<double,double> getB(double m0, double m1, double m2, double m3) {
    double b1 = (m1 * m2 - m0 * m3) / (m0 * m2 - m1 * m1);
    double b2 = (m2 + m1 * b1) / -m0;
    return make_pair(b1, b2);
}

pair<double, double> getRoot(double b1, double b2) {
    double p1 = (-b1 + sqrt(b1 * b1 - 4 * b2))/(2 * b2);
    double p2 = (-b1 - sqrt(b1 * b1 - 4 * b2))/(2 * b2);
    return make_pair(p1, p2);
}

pair<double,double> getK(double p1, double p2, double m0, double m1) {
    double a, b, c, d, e, f;
    a = pow(p1,-1);
    b = pow(p2,-1);
    c = pow(p1,-2);
    d = pow(p2,-2);
    e = -m0;
    f = -m1;
    double k2 = (a*f - c*e) / (a*d - b*c);
    double k1 = (e - b * k2) / a;
    return make_pair(k1, k2);
}

int main() {
    double C0 = 400e-15;
    double C1 = 1.4e-12;
    double C2 = 20e-15;
    double C3 = 20e-15;
    double R0 = 100;
    double R1 = 40;
    double R2 = 80;
    double R3 = 20;
    
    double D0 = R0 * (C0 + C1 + C2 + C3);
    double D1 = D0 + R1 * (C1 + C2 + C3);
    double D2 = D1 + R2 * C2;
    double D3 = D1 + R3 * C3;

    // [node #][order #]
    double m00, m10, m20, m30;
    double m01, m11, m21, m31;
    double m02, m12, m22, m32;
    double m03, m13, m23, m33;
    
    // Calculate m0
    m00 = m10 = m20 = m30 = 1;

    // Calculate m1
    m01 =   0 - R0 * (m00 * C0 + m10 * C1 + m20 * C2 + m30 * C3);
    m11 = m01 - R1 * (m10 * C1 + m20 * C2 + m30 * C3);
    m21 = m11 - R2 * (m20 * C2);
    m31 = m11 - R3 * (m30 * C3);

    // Calculate m2
    m02 =   0 - R0 * (m01 * C0 + m11 * C1 + m21 * C2 + m31 * C3);
    m12 = m02 - R1 * (m11 * C1 + m21 * C2 + m31 * C3);
    m22 = m12 - R2 * (m21 * C2);
    m32 = m12 - R3 * (m31 * C3);

    // Calculate m3
    m03 =   0 - R0 * (m01 * C0 + m12 * C1 + m22 * C2 + m32 * C3);
    m13 = m03 - R1 * (m12 * C1 + m22 * C2 + m32 * C3);
    m23 = m13 - R2 * (m22 * C2);
    m33 = m13 - R3 * (m32 * C3);

    auto[b01, b02] = getB(m00, m01, m02, m03);
    auto[b11, b12] = getB(m10, m11, m12, m13);
    auto[b21, b22] = getB(m20, m21, m22, m23);
    auto[b31, b32] = getB(m30, m31, m32, m33);

    auto[p01, p02] = getRoot(b01, b02);
    auto[p11, p12] = getRoot(b11, b12);
    auto[p21, p22] = getRoot(b21, b22);
    auto[p31, p32] = getRoot(b31, b32);

    auto[k01, k02] = getK(p01, p02, m00, m01);
    auto[k11, k12] = getK(p11, p12, m10, m11);
    auto[k21, k22] = getK(p21, p22, m20, m21);
    auto[k31, k32] = getK(p31, p32, m30, m31);

    printf("D0 = %.6e\t", D0);
    printf("D1 = %.6e\t", D1);
    printf("D2 = %.6e\t", D2);
    printf("D3 = %.6e\n\n", D3);
    
    printf("m00 = %.6e\t", m00);
    printf("m10 = %.6e\t", m10);
    printf("m20 = %.6e\t", m20);
    printf("m30 = %.6e\n", m30);
    printf("m01 = %.6e\t", m01);
    printf("m11 = %.6e\t", m11);
    printf("m21 = %.6e\t", m21);
    printf("m31 = %.6e\n", m31);
    printf("m02 = %.6e\t", m02);
    printf("m12 = %.6e\t", m12);
    printf("m22 = %.6e\t", m22);
    printf("m32 = %.6e\n", m32);
    printf("m03 = %.6e\t", m03);
    printf("m13 = %.6e\t", m13);
    printf("m23 = %.6e\t", m23);
    printf("m33 = %.6e\n\n", m33);

    printf("b01 = %.6e\t", b01);
    printf("b11 = %.6e\t", b11);
    printf("b21 = %.6e\t", b21);
    printf("b31 = %.6e\n", b31);
    printf("b02 = %.6e\t", b02);
    printf("b12 = %.6e\t", b12);
    printf("b22 = %.6e\t", b22);
    printf("b32 = %.6e\n\n", b32);

    printf("p01 = %.6e\t", p01);
    printf("p11 = %.6e\t", p11);
    printf("p21 = %.6e\t", p21);
    printf("p31 = %.6e\n", p31);
    printf("p02 = %.6e\t", p02);
    printf("p12 = %.6e\t", p12);
    printf("p22 = %.6e\t", p22);
    printf("p32 = %.6e\n\n", p32);

    printf("k01 = %.6e\t", k01);
    printf("k11 = %.6e\t", k11);
    printf("k21 = %.6e\t", k21);
    printf("k31 = %.6e\n", k31);
    printf("k02 = %.6e\t", k02);
    printf("k12 = %.6e\t", k12);
    printf("k22 = %.6e\t", k22);
    printf("k32 = %.6e\n\n", k32);


    return 0;
}

File: HW16/HW16.m
clear; clc;
C0 = 400e-15;
C1 = 1.4e-12;
C2 = 20e-15;
C3 = 20e-15;
R0 = 100;
R1 = 40;
R2 = 80;
R3 = 20;
D0 = R0 * (C0 + C1 + C2 + C3);
D1 = D0 + R1 * (C1 + C2 + C3);
D2 = D1 + R2 * C2;
D3 = D1 + R3 * C3;
Elmore = [D0; D1; D2; D3]

% Calculate m0
m00 = 1;
m10 = 1;
m20 = 1;
m30 = 1;

% Calculate m1
m01 =   0 - R0 * (m00 * C0 + m10 * C1 + m20 * C2 + m30 * C3);
m11 = m01 - R1 * (m10 * C1 + m20 * C2 + m30 * C3);
m21 = m11 - R2 * (m20 * C2);
m31 = m11 - R3 * (m30 * C3);

% Calculate m2
m02 = vpa(  0 - R0 * (m01 * C0 + m11 * C1 + m21 * C2 + m31 * C3));
m12 = vpa(m02 - R1 * (m11 * C1 + m21 * C2 + m31 * C3));
m22 = vpa(m12 - R2 * (m21 * C2));
m32 = vpa(m12 - R3 * (m31 * C3));

% Calculate m3
m03 = vpa(  0 - R0 * (m01 * C0 + m12 * C1 + m22 * C2 + m32 * C3));
m13 = vpa(m03 - R1 * (m12 * C1 + m22 * C2 + m32 * C3));
m23 = vpa(m13 - R2 * (m22 * C2));
m33 = vpa(m13 - R3 * (m32 * C3));

A0 = [m00 m01; m01 m02];
A1 = [m10 m11; m11 m12];
A2 = [m20 m21; m21 m22];
A3 = [m30 m31; m31 m32];
Y0 = [-m02; -m03];
Y1 = [-m12; -m13];
Y2 = [-m22; -m23];
Y3 = [-m32; -m33];
B0 = linsolve(A0, Y0);
B1 = linsolve(A1, Y1);
B2 = linsolve(A2, Y2);
B3 = linsolve(A3, Y3);

P0 = roots([B0(1) B0(2) 1]);
P1 = roots([B1(1) B1(2) 1]);
P2 = roots([B2(1) B2(2) 1]);
P3 = roots([B3(1) B3(2) 1]);

A0 = [P0(1)^-1 P0(2)^-1; P0(1)^-2 P0(2)^-2];
A1 = [P1(1)^-1 P1(2)^-1; P1(1)^-2 P1(2)^-2];
A2 = [P2(1)^-1 P2(2)^-1; P2(1)^-2 P2(2)^-2];
A3 = [P3(1)^-1 P3(2)^-1; P3(1)^-2 P3(2)^-2];
Y0 = [-m00; -m01];
Y1 = [-m10; -m11];
Y2 = [-m20; -m21];
Y3 = [-m30; -m31];
K0 = linsolve(A0, Y0);
K1 = linsolve(A1, Y1);
K2 = linsolve(A2, Y2);
K3 = linsolve(A3, Y3);

syms s t
V0 = sum(K0./P0)*(-1/s) + (K0(1)/P0(1))*(1/(s-P0(1))) + (K0(2)/P0(2))*(1/(s-P0(2)));
V1 = sum(K1./P1)*(-1/s) + (K1(1)/P1(1))*(1/(s-P1(1))) + (K1(2)/P1(2))*(1/(s-P1(2)));
V2 = sum(K2./P2)*(-1/s) + (K2(1)/P2(1))*(1/(s-P2(1))) + (K2(2)/P2(2))*(1/(s-P2(2)));
V3 = sum(K3./P3)*(-1/s) + (K3(1)/P3(1))*(1/(s-P3(1))) + (K3(2)/P3(2))*(1/(s-P3(2)));

V0 = ilaplace(V0);
V1 = ilaplace(V1);
V2 = ilaplace(V2);
V3 = ilaplace(V3);

AWE0 = solve(V0 == 0.5, t);
AWE1 = solve(V1 == 0.5, t);
AWE2 = solve(V2 == 0.5, t);
AWE3 = solve(V3 == 0.5, t);
AWE = [AWE0; AWE1; AWE2; AWE3]

delayU = 1.72206e-10;
delayV = 1.71019e-10;
PEU_Elm = 100*abs(D2 - delayU)/delayU;
PEV_Elm = 100*abs(D3 - delayV)/delayV;
PEU_AWE = 100*abs(AWE2 - delayU)/delayU;
PEV_AWE = 100*abs(AWE3 - delayV)/delayV;
fprintf("Percent Errors:\n")
fprintf("\t\tSink U\t\tSink V\n")
fprintf("Elmore:\t%f%%\t%f%%\n", PEU_Elm, PEV_Elm)
fprintf("AWE:\t%f%%\t%f%%\n", PEU_AWE, PEV_AWE)